### ISSUE TITLE
Add support for aria-labelledby and aria-describedby

### ISSUE THREAD

**Issue by jeremyf** [heartcombo/simple_form#1175]
In using SimpleForm we are often finding that we need aria-labelledby and aria-describedby; I have a rudimentary solution in place but it looks very fragile and subject to breakage as SimpleForm moves forward.

I'm exploring adding a new component (akin to the `SimpleForm::Components::HTML5`) for `SimpleForm::Components::AriaLabelledBy` and `SimpleForm::Components::AriaDescribedBy`.

I'm presently looking at how best to go about submitting a pull request for this behavior. Its looking like a rather lengthy process as I orient to the code of SimpleForm. I want to make sure that this is something that would be a reasonable addition to the SimpleForm ecosystem.

**Comment by jmuheim** [comment]
Is there any progress in this? I'm facing the task to make Rails forms more accessible which needs to assign the validation errors to the specific input fields so screenreader users see them. Also, the focus should be set into the first invalid input field.

At the time being, I'm simply using a JavaScript workaround to do this:

```coffee
class App.FormErrorsAccessibilizer
  constructor: (el) ->
    @$el = $(el) # Always pass an element to the constructor and make it available as a jQuery selector!
    @formGroups = @$el.find('.form-group.has-error')

    if @formGroups.size() > 0
      @describe_inputs_with_help_blocks()
      @focus_first_input_with_help_block()

  describe_inputs_with_help_blocks: ->
    for formGroup in @formGroups
      $formGroup = $(formGroup)

      $help  = $formGroup.find('.help-block')
      $input = $formGroup.find(':input')

      input_id = $input.attr('id')
      help_id  = "#{input_id}_help"

      $help.attr('id', help_id)
      $input.attr('aria-describedby', help_id)

  focus_first_input_with_help_block: ->
    @formGroups.first().find(':input').focus()
```

This works nicely, but it's a bit error-prone and hard to test. It would be very nice if Simple_Form would offer some accessibility plugin.

**Comment by jmuheim** [comment]
I just had another idea, something like fixing this problem in the configuration:

```ruby
      ba.use :input, aria: {describedby: 'xyz_error xyz_hint'} # xyz would be the input's ID
      ba.use :error, wrap_with: { tag: 'span', class: 'help-block', id: 'xyz_error' }
      ba.use :hint,  wrap_with: { tag: 'p', class: 'help-block', id: 'xyz_hint' }
```

The tricky part here is that `aria-describedby` should only assigned to the input when there's an alert and/or a hint. Any idea on how to accomplish this? One could maybe monkey-patch the `Input` class, but I don't know whether this is the best approach.

**Comment by jeremyf** [comment]
@jmuheim I believe the changes I proposed might work for what you need. But I've had to set them aside. Keep me posted if you dive into this.

**Comment by sarahjones** [comment]
Have there been any developments on this topic?  We're in the same boat, where something like https://github.com/plataformatec/simple_form/issues/1175#issuecomment-86887671 would be awesome.  Thanks!

**Comment by mattwr18** [comment]
For others finding this issue later, I got this working by using:
```
input_html: { aria: { labelledby: 'your-label' } }
```
maybe this is obvious for some, but it took a bit of time to get it right.

**Comment by javinto** [comment]
You might like this solution as a SimpleForm component:

Save this file into /config/initializers/simple_form/components/aria.rb

Use this one like so in your wrappers (before b.use :label and b.use :input)
b.use :aria

# frozen_string_literal: true
```
module SimpleForm
  module Components

    module Aria
      def aria(wrapper_options = nil)

        # Optionally add an aria_scope to the label and hint ID's to prevent any duplicate ids within the whole page
        if @aria_scope= options[:aria_scope].presence or @aria_scope= @builder.options[:html].dig(:aria_scope)
          @aria_scope+= '_'
        end

        # Actually we don't need this one as <Label for='input_id'> already labels the input field. But in case one does not use Label tags it's helpfull.
        if options[:label]!=false and label_id= aria_dom_id('label')
          @options[:label_html]||={}
          @options[:label_html][:id]||= label_id
          @input_html_options[:'aria-labelledby']= @options[:label_html][:id]
        end

        # Add aria-describedby to Hint component
        if has_hint? and hint_id= aria_dom_id('hint')
          @options[:hint_html]||={}
          @options[:hint_html][:id]||= hint_id
          @input_html_options[:'aria-describedby']= @options[:hint_html][:id]
        end

        @aria||= ''
      end

      def has_aria?
        options[:aria] != false
      end


      private

      def aria_dom_id(suffix)
        "#{@aria_scope}#{object_name}_#{attribute_name}_#{suffix}"
      end

    end
  end
end

SimpleForm::Inputs::Base.send(:include, SimpleForm::Components::Aria)
```
